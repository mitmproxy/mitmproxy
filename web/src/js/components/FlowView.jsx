import React, { Component } from 'react'
import { connect } from 'react-redux'
import _ from 'lodash'

import Nav from './FlowView/Nav'
import { ErrorView as Error, Request, Response } from './FlowView/Messages'
import Details from './FlowView/Details'

import { selectTab } from '../ducks/ui/flow'

export const allTabs = { Request, Response, Error, Details }

function FlowView({ flow, tabName, selectTab }) {

    // only display available tab names
    const tabs = ['request', 'response', 'error'].filter(k => flow[k])
    tabs.push("details")

    if (tabs.indexOf(tabName) < 0) {
        if (tabName === 'response' && flow.error) {
            tabName = 'error'
        } else if (tabName === 'error' && flow.response) {
            tabName = 'response'
        } else {
            tabName = tabs[0]
        }
    }

    const Tab = allTabs[_.capitalize(tabName)]

    return (
        <div className="flow-detail">
            <Nav
                tabs={tabs}
                active={tabName}
                onSelectTab={selectTab}
            />
            <Tab flow={flow}/>
        </div>
    )
}

export default connect(
    state => ({
        flow: state.flows.byId[state.flows.selected[0]],
        tabName: state.ui.flow.tab,
    }),
    {
        selectTab,
    }
)(FlowView)
