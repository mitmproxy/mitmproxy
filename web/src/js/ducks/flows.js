import { fetchApi } from "../utils"
import reduceStore from "./utils/store"
import * as storeActions from "./utils/store"
import Filt from "../filt/filt"
import { RequestUtils } from "../flow/utils"

export const ADD            = 'FLOWS_ADD'
export const UPDATE         = 'FLOWS_UPDATE'
export const REMOVE         = 'FLOWS_REMOVE'
export const RECEIVE        = 'FLOWS_RECEIVE'
export const SELECT         = 'FLOWS_SELECT'
export const SET_FILTER     = 'FLOWS_SET_FILTER'
export const SET_SORT       = 'FLOWS_SET_SORT'
export const SET_HIGHLIGHT  = 'FLOWS_SET_HIGHLIGHT'
export const REQUEST_ACTION = 'FLOWS_REQUEST_ACTION'


const defaultState = {
    highlight: null,
    filter: null,
    sort: { column: null, desc: false },
    selected: [],
    ...reduceStore(undefined, {})
}

export default function reduce(state = defaultState, action) {
    switch (action.type) {

        case ADD:
        case UPDATE:
        case REMOVE:
        case RECEIVE:
            let storeAction = storeActions[action.cmd](
                action.data,
                makeFilter(state.filter),
                makeSort(state.sort)
            )

            let selected = state.selected
            if(action.type === REMOVE && state.selected.includes(action.data)) {
                if(state.selected.length > 1){
                    selected = selected.filter(x => x !== action.data)
                } else {
                    selected = []
                    if (action.data in state.viewIndex && state.view.length > 1) {
                        let currentIndex = state.viewIndex[action.data],
                            nextSelection
                        if(currentIndex === state.view.length -1){ // last row
                            nextSelection = state.view[currentIndex - 1]
                        } else {
                            nextSelection = state.view[currentIndex + 1]
                        }
                        selected.push(nextSelection.id)
                    }
                }
            }

            return {
                ...state,
                selected,
                ...reduceStore(state, storeAction)
            }

        case SET_FILTER:
            return {
                ...state,
                filter: action.filter,
                ...reduceStore(state, storeActions.setFilter(makeFilter(action.filter), makeSort(state.sort)))
            }

        case SET_HIGHLIGHT:
            return {
                ...state,
                highlight: action.highlight
            }

        case SET_SORT:
            return {
                ...state,
                sort: action.sort,
                ...reduceStore(state, storeActions.setSort(makeSort(action.sort)))
            }

        case SELECT:
            return {
                ...state,
                selected: action.flowIds
            }

        default:
            return state
    }
}


const sortKeyFuns = {

    TLSColumn: flow => flow.request.scheme,

    PathColumn: flow => RequestUtils.pretty_url(flow.request),

    MethodColumn: flow => flow.request.method,

    StatusColumn: flow => flow.response && flow.response.status_code,

    TimeColumn: flow => flow.response && flow.response.timestamp_end - flow.request.timestamp_start,

    SizeColumn: flow => {
        let total = flow.request.contentLength
        if (flow.response) {
            total += flow.response.contentLength || 0
        }
        return total
    },
}

export function makeFilter(filter) {
    if (!filter) {
        return
    }
    return Filt.parse(filter)
}

export function makeSort({ column, desc }) {
    const sortKeyFun = sortKeyFuns[column]
    if (!sortKeyFun) {
        return
    }
    return (a, b) => {
        const ka = sortKeyFun(a)
        const kb = sortKeyFun(b)
        if (ka > kb) {
            return desc ? -1 : 1
        }
        if (ka < kb) {
            return desc ? 1 : -1
        }
        return 0
    }
}

export function setFilter(filter) {
    return { type: SET_FILTER, filter }
}

export function setHighlight(highlight) {
    return { type: SET_HIGHLIGHT, highlight }
}

export function setSort(column, desc) {
    return { type: SET_SORT, sort: { column, desc } }
}

export function selectRelative(flows, shift) {
    let currentSelectionIndex = flows.viewIndex[flows.selected[0]]
    let minIndex = 0
    let maxIndex = flows.view.length - 1
    let newIndex
    if (currentSelectionIndex === undefined) {
        newIndex = (shift < 0) ? minIndex : maxIndex
    } else {
        newIndex = currentSelectionIndex + shift
        newIndex = window.Math.max(newIndex, minIndex)
        newIndex = window.Math.min(newIndex, maxIndex)
    }
    let flow = flows.view[newIndex]
    return select(flow ? flow.id : undefined)
}


export function resume(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/resume`, { method: 'POST' })
}

export function resumeAll() {
    return dispatch => fetchApi('/flows/resume', { method: 'POST' })
}

export function kill(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/kill`, { method: 'POST' })
}

export function killAll() {
    return dispatch => fetchApi('/flows/kill', { method: 'POST' })
}


export function remove(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}`, { method: 'DELETE' })
}

export function duplicate(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/duplicate`, { method: 'POST' })
}

export function replay(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/replay`, { method: 'POST' })
}

export function revert(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/revert`, { method: 'POST' })
}

export function update(flow, data) {
    return dispatch => fetchApi.put(`/flows/${flow.id}`, data)
}

export function uploadContent(flow, file, type) {
    const body = new FormData()
    file       = new window.Blob([file], { type: 'plain/text' })
    body.append('file', file)
    return dispatch => fetchApi(`/flows/${flow.id}/${type}/content.data`, { method: 'POST', body })
}


export function clear() {
    return dispatch => fetchApi('/clear', { method: 'POST' })
}

export function download() {
    window.location = '/flows/dump'
    return { type: REQUEST_ACTION }
}

export function upload(file) {
    const body = new FormData()
    body.append('file', file)
    return dispatch => fetchApi('/flows/dump', { method: 'POST', body })
}


export function select(id) {
    return {
        type: SELECT,
        flowIds: id ? [id] : []
    }
}
