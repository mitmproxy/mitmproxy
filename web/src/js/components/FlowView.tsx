import * as React from "react"
import {FunctionComponent} from "react"
import {Request, Response} from './FlowView/HttpMessages'
import {Request as DnsRequest, Response as DnsResponse} from './FlowView/DnsMessages'
import Connection from './FlowView/Connection'
import Error from "./FlowView/Error"
import Timing from "./FlowView/Timing"
import WebSocket from "./FlowView/WebSocket"

import {selectTab} from '../ducks/ui/flow'
import {useAppDispatch, useAppSelector} from "../ducks";
import {Flow} from "../flow";
import classnames from "classnames";
import TcpMessages from "./FlowView/TcpMessages";

type TabProps = {
    flow: Flow
}

export const allTabs: { [name: string]: FunctionComponent<TabProps> & { displayName: string } } = {
    request: Request,
    response: Response,
    error: Error,
    connection: Connection,
    timing: Timing,
    websocket: WebSocket,
    messages: TcpMessages,
    dnsrequest: DnsRequest,
    dnsresponse: DnsResponse,
}

export function tabsForFlow(flow: Flow): string[] {
    let tabs;
    switch (flow.type) {
        case "http":
            tabs = ['request', 'response', 'websocket'].filter(k => flow[k])
            break
        case "tcp":
            tabs = ["messages"]
            break
        case "dns":
            tabs = ['request', 'response'].filter(k => flow[k]).map(s => "dns" + s)
            break
    }

    if (flow.error)
        tabs.push("error")
    tabs.push("connection")
    tabs.push("timing")
    return tabs;
}

export default function FlowView() {
    const dispatch = useAppDispatch(),
        flow = useAppSelector(state => state.flows.byId[state.flows.selected[0]]),
        tabs = tabsForFlow(flow);

    let active = useAppSelector(state => state.ui.flow.tab)
    if (tabs.indexOf(active) < 0) {
        if (active === 'response' && flow.error) {
            active = 'error'
        } else if (active === 'error' && "response" in flow) {
            active = 'response'
        } else {
            active = tabs[0]
        }
    }
    const Tab = allTabs[active];

    return (
        <div className="flow-detail">
            <nav className="nav-tabs nav-tabs-sm">
                {tabs.map(tabId => (
                    <a key={tabId} href="#" className={classnames({active: active === tabId})}
                       onClick={event => {
                           event.preventDefault()
                           dispatch(selectTab(tabId))
                       }}>
                        {allTabs[tabId].displayName}
                    </a>
                ))}
            </nav>
            <Tab flow={flow}/>
        </div>
    )
}
