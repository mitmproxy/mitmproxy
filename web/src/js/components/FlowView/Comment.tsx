import { Flow } from "../../flow";
import * as React from "react";
import ValueEditor from "../editors/ValueEditor";
import { useAppDispatch } from "../../ducks";
import * as flowActions from "../../ducks/flows";

export default function Comment({ flow }: { flow: Flow }) {
    const dispatch = useAppDispatch();

    return (
        <section className="timing">
            <h4>Comment</h4>
            <ValueEditor
                className="kv-value"
                content={flow.comment}
                onEditDone={(comment) => {
                    dispatch(flowActions.update(flow, { comment }));
                }}
                placeholder="empty"
            />
        </section>
    );
}
Comment.displayName = "Comment";
