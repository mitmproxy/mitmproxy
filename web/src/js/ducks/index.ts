import {
    AnyAction,
    applyMiddleware,
    combineReducers,
    compose,
    createStore as createReduxStore,
    PreloadedState,
    Store
} from "redux"
import eventLog from "./eventLog"
import flows from "./flows"
import ui from "./ui/index"
import connection from "./connection"
import options from './options'
import commandBar from "./commandBar";
import thunk, {ThunkAction, ThunkDispatch, ThunkMiddleware} from "redux-thunk";
import {TypedUseSelectorHook, useDispatch, useSelector} from "react-redux";
import backendState from "./backendState";
import options_meta from "./options_meta";


// @ts-ignore
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const rootReducer = combineReducers({
    commandBar,
    eventLog,
    flows,
    connection,
    ui,
    options,
    options_meta,
    backendState,
});
export type RootState = ReturnType<typeof rootReducer>


export type AppThunk<ReturnType = void> = ThunkAction<ReturnType,
    RootState,
    unknown,
    AnyAction>

export const createAppStore = (preloadedState?: PreloadedState<RootState>): Store<RootState> => {
    return createReduxStore(
        rootReducer,
        preloadedState,
        composeEnhancers(applyMiddleware(
            thunk as ThunkMiddleware<RootState>
        )))
};

export const store = createAppStore(undefined);

// this would be correct, but PyCharm bails on it
// export type AppDispatch = typeof store.dispatch
// instead:
export type AppDispatch = ThunkDispatch<RootState, void, AnyAction>
export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
