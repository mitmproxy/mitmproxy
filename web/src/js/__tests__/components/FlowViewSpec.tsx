import * as React from "react"
import {render, screen} from "../test-utils";
import FlowView from "../../components/FlowView";
import * as flowActions from "../../ducks/flows"
import fetchMock, {enableFetchMocks} from "jest-fetch-mock";
import {fireEvent} from "@testing-library/react";

enableFetchMocks();

test("FlowView", async () => {
    fetchMock.mockReject(new Error("backend missing"));

    const {asFragment, store} = render(<FlowView/>);
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText("Response"));
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText("WebSocket"));
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText("Connection"));
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText("Timing"));
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText("Error"));
    expect(asFragment()).toMatchSnapshot();

    store.dispatch(flowActions.select(store.getState().flows.list[2].id));

    fireEvent.click(screen.getByText("TCP Messages"));
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText("Error"));
    expect(asFragment()).toMatchSnapshot();

    store.dispatch(flowActions.select(store.getState().flows.list[3].id));

    fireEvent.click(screen.getByText("Request"));
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText("Response"));
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText("Error"));
    expect(asFragment()).toMatchSnapshot();
});
