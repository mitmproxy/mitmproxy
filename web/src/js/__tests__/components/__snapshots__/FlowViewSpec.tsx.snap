// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowView 1`] = `
<DocumentFragment>
  <div
    class="flow-detail"
  >
    <nav
      class="nav-tabs nav-tabs-sm"
    >
      <a
        class="active"
        href="#"
      >
        Request
      </a>
      <a
        class=""
        href="#"
      >
        Response
      </a>
      <a
        class=""
        href="#"
      >
        WebSocket
      </a>
      <a
        class=""
        href="#"
      >
        Error
      </a>
      <a
        class=""
        href="#"
      >
        Connection
      </a>
      <a
        class=""
        href="#"
      >
        Timing
      </a>
    </nav>
    <section
      class="request"
    >
      <div
        class="first-line request-line"
      >
        <div>
          <span
            class="inline-input has-success"
            tabindex="0"
          >
            GET
          </span>
           
          <span
            class="inline-input has-success"
            tabindex="0"
          >
            http://address:22/second
          </span>
           
          <span
            class="inline-input has-success"
            tabindex="0"
          >
            HTTP/1.1
          </span>
        </div>
      </div>
      <div
        class="kv-editor headers"
      >
        <div
          class="kv-row"
        >
          <span
            class="inline-input kv-key"
            tabindex="0"
          >
            header
          </span>
          : 
          <span
            class="inline-input kv-value"
            placeholder="empty"
            tabindex="0"
          >
            qvalue
          </span>
        </div>
        <div
          class="kv-row"
        >
          <span
            class="inline-input kv-key"
            tabindex="0"
          >
            content-length
          </span>
          : 
          <span
            class="inline-input kv-value"
            placeholder="empty"
            tabindex="0"
          >
            7
          </span>
        </div>
        <div
          aria-label="Add"
          class="kv-add-row fa fa-plus-square-o"
          role="button"
        />
      </div>
      <hr />
      <div
        class="contentview"
      >
        <div
          class="controls"
        >
          <h5>
            Loading...
          </h5>
          <button
            class="btn-xs btn btn-default"
          >
            <i
              class="fa fa-edit"
            />
             Edit
          </button>
           
          <a
            class="btn btn-default btn-xs"
            href="#"
            title="Upload a file to replace the content."
          >
            <i
              class="fa fa-fw fa-upload"
            />
            Replace
            <input
              class="hidden"
              type="file"
            />
          </a>
           
          <a
            class="btn btn-default btn-xs"
            href="#"
          >
            <span>
              <i
                class="fa fa-fw fa-files-o"
              />
               
              <b>
                View:
              </b>
               auto 
              <span
                class="caret"
              />
            </span>
          </a>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;

exports[`FlowView 2`] = `
<DocumentFragment>
  <div
    class="flow-detail"
  >
    <nav
      class="nav-tabs nav-tabs-sm"
    >
      <a
        class=""
        href="#"
      >
        Request
      </a>
      <a
        class="active"
        href="#"
      >
        Response
      </a>
      <a
        class=""
        href="#"
      >
        WebSocket
      </a>
      <a
        class=""
        href="#"
      >
        Error
      </a>
      <a
        class=""
        href="#"
      >
        Connection
      </a>
      <a
        class=""
        href="#"
      >
        Timing
      </a>
    </nav>
    <section
      class="response"
    >
      <div
        class="first-line response-line"
      >
        <span
          class="inline-input has-success"
          tabindex="0"
        >
          HTTP/1.1
        </span>
         
        <span
          class="inline-input has-success"
          tabindex="0"
        >
          200
        </span>
         
        <span
          class="inline-input"
          tabindex="0"
        >
          OK
        </span>
      </div>
      <div
        class="kv-editor headers"
      >
        <div
          class="kv-row"
        >
          <span
            class="inline-input kv-key"
            tabindex="0"
          >
            header-response
          </span>
          : 
          <span
            class="inline-input kv-value"
            placeholder="empty"
            tabindex="0"
          >
            svalue
          </span>
        </div>
        <div
          class="kv-row"
        >
          <span
            class="inline-input kv-key"
            tabindex="0"
          >
            content-length
          </span>
          : 
          <span
            class="inline-input kv-value"
            placeholder="empty"
            tabindex="0"
          >
            7
          </span>
        </div>
        <div
          aria-label="Add"
          class="kv-add-row fa fa-plus-square-o"
          role="button"
        />
      </div>
      <hr />
      <div
        class="contentview"
      >
        <div
          class="controls"
        >
          <h5>
            Loading...
          </h5>
          <button
            class="btn-xs btn btn-default"
          >
            <i
              class="fa fa-edit"
            />
             Edit
          </button>
           
          <a
            class="btn btn-default btn-xs"
            href="#"
            title="Upload a file to replace the content."
          >
            <i
              class="fa fa-fw fa-upload"
            />
            Replace
            <input
              class="hidden"
              type="file"
            />
          </a>
           
          <a
            class="btn btn-default btn-xs"
            href="#"
          >
            <span>
              <i
                class="fa fa-fw fa-files-o"
              />
               
              <b>
                View:
              </b>
               auto 
              <span
                class="caret"
              />
            </span>
          </a>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;

exports[`FlowView 3`] = `
<DocumentFragment>
  <div
    class="flow-detail"
  >
    <nav
      class="nav-tabs nav-tabs-sm"
    >
      <a
        class=""
        href="#"
      >
        Request
      </a>
      <a
        class=""
        href="#"
      >
        Response
      </a>
      <a
        class="active"
        href="#"
      >
        WebSocket
      </a>
      <a
        class=""
        href="#"
      >
        Error
      </a>
      <a
        class=""
        href="#"
      >
        Connection
      </a>
      <a
        class=""
        href="#"
      >
        Timing
      </a>
    </nav>
    <section
      class="websocket"
    >
      <h4>
        WebSocket
      </h4>
      <div
        class="contentview"
      >
        <div
          class="controls"
        >
          <h5>
            3 Messages
          </h5>
          <a
            class="btn btn-default btn-xs"
            href="#"
          >
            <span>
              <i
                class="fa fa-fw fa-files-o"
              />
               
              <b>
                View:
              </b>
               auto 
              <span
                class="caret"
              />
            </span>
          </a>
        </div>
      </div>
      <div>
        <i
          class="fa fa-fw fa-window-close text-muted"
        />
          Closed by server with code 1000 (Close Reason).
        <small
          class="pull-right"
        >
          1999-12-31 23:00:05.000
        </small>
      </div>
    </section>
  </div>
</DocumentFragment>
`;

exports[`FlowView 4`] = `
<DocumentFragment>
  <div
    class="flow-detail"
  >
    <nav
      class="nav-tabs nav-tabs-sm"
    >
      <a
        class=""
        href="#"
      >
        Request
      </a>
      <a
        class=""
        href="#"
      >
        Response
      </a>
      <a
        class=""
        href="#"
      >
        WebSocket
      </a>
      <a
        class=""
        href="#"
      >
        Error
      </a>
      <a
        class="active"
        href="#"
      >
        Connection
      </a>
      <a
        class=""
        href="#"
      >
        Timing
      </a>
    </nav>
    <section
      class="detail"
    >
      <h4>
        Client Connection
      </h4>
      <table
        class="connection-table"
      >
        <tbody>
          <tr>
            <td>
              Address:
            </td>
            <td>
              127.0.0.1:22
            </td>
          </tr>
          <tr>
            <td>
              <abbr
                title="TLS Server Name Indication"
              >
                SNI
              </abbr>
              :
            </td>
            <td>
              address
            </td>
          </tr>
          <tr>
            <td>
              <abbr
                title="ALPN protocol negotiated"
              >
                ALPN
              </abbr>
              :
            </td>
            <td>
              http/1.1
            </td>
          </tr>
          <tr>
            <td>
              TLS Version:
            </td>
            <td>
              TLSv1.2
            </td>
          </tr>
          <tr>
            <td>
              TLS Cipher:
            </td>
            <td>
              cipher
            </td>
          </tr>
        </tbody>
      </table>
      <h4>
        Server Connection
      </h4>
      <table
        class="connection-table"
      >
        <tbody>
          <tr>
            <td>
              Address:
            </td>
            <td>
              address:22
            </td>
          </tr>
          <tr>
            <td>
              Resolved address:
            </td>
            <td>
              192.168.0.1:22
            </td>
          </tr>
          <tr>
            <td>
              Source address:
            </td>
            <td>
              address:22
            </td>
          </tr>
          <tr>
            <td>
              <abbr
                title="TLS Server Name Indication"
              >
                SNI
              </abbr>
              :
            </td>
            <td>
              address
            </td>
          </tr>
          <tr>
            <td>
              TLS Version:
            </td>
            <td>
              TLSv1.2
            </td>
          </tr>
        </tbody>
      </table>
      <h4>
        Server Certificate
      </h4>
      <table
        class="certificate-table"
      >
        <tbody>
          <tr>
            <td>
              Type
            </td>
            <td>
              RSA, 2048 bits
            </td>
          </tr>
          <tr>
            <td>
              SHA256 digest
            </td>
            <td>
              e5f62a1175031b6feb959bc8e6dd0f8e2546dbbf7c32da39534309d8aa92967c
            </td>
          </tr>
          <tr>
            <td>
              Valid from
            </td>
            <td>
              2020-11-03 07:03:27
            </td>
          </tr>
          <tr>
            <td>
              Valid to
            </td>
            <td>
              2040-10-29 07:03:27
            </td>
          </tr>
          <tr>
            <td>
              Subject Alternative Names
            </td>
            <td>
              example.mitmproxy.org
            </td>
          </tr>
          <tr>
            <td>
              Subject
            </td>
            <td>
              <dl
                class="cert-attributes"
              >
                <dt>
                  C
                </dt>
                <dd>
                  AU
                </dd>
                <dt>
                  ST
                </dt>
                <dd>
                  Some-State
                </dd>
                <dt>
                  O
                </dt>
                <dd>
                  Internet Widgits Pty Ltd
                </dd>
              </dl>
            </td>
          </tr>
          <tr>
            <td>
              Issuer
            </td>
            <td>
              <dl
                class="cert-attributes"
              >
                <dt>
                  C
                </dt>
                <dd>
                  AU
                </dd>
                <dt>
                  ST
                </dt>
                <dd>
                  Some-State
                </dd>
                <dt>
                  O
                </dt>
                <dd>
                  Internet Widgits Pty Ltd
                </dd>
              </dl>
            </td>
          </tr>
          <tr>
            <td>
              Serial
            </td>
            <td>
              247170098335718583458667965517443538258472437317
            </td>
          </tr>
        </tbody>
      </table>
    </section>
  </div>
</DocumentFragment>
`;

exports[`FlowView 5`] = `
<DocumentFragment>
  <div
    class="flow-detail"
  >
    <nav
      class="nav-tabs nav-tabs-sm"
    >
      <a
        class=""
        href="#"
      >
        Request
      </a>
      <a
        class=""
        href="#"
      >
        Response
      </a>
      <a
        class=""
        href="#"
      >
        WebSocket
      </a>
      <a
        class=""
        href="#"
      >
        Error
      </a>
      <a
        class=""
        href="#"
      >
        Connection
      </a>
      <a
        class="active"
        href="#"
      >
        Timing
      </a>
    </nav>
    <section
      class="timing"
    >
      <h4>
        Timing
      </h4>
      <table
        class="timing-table"
      >
        <tbody>
          <tr>
            <td>
              Client conn. established:
            </td>
            <td>
              1999-12-31 23:00:00.000
              <span
                class="text-muted"
              >
                (0ms)
              </span>
            </td>
          </tr>
          <tr>
            <td>
              First request byte:
            </td>
            <td>
              1999-12-31 23:00:00.000
            </td>
          </tr>
          <tr>
            <td>
              Client conn. TLS handshake:
            </td>
            <td>
              1999-12-31 23:00:01.000
              <span
                class="text-muted"
              >
                (1s)
              </span>
            </td>
          </tr>
          <tr>
            <td>
              Request complete:
            </td>
            <td>
              1999-12-31 23:00:01.000
              <span
                class="text-muted"
              >
                (1s)
              </span>
            </td>
          </tr>
          <tr>
            <td>
              Server conn. initiated:
            </td>
            <td>
              1999-12-31 23:00:02.000
              <span
                class="text-muted"
              >
                (2s)
              </span>
            </td>
          </tr>
          <tr>
            <td>
              First response byte:
            </td>
            <td>
              1999-12-31 23:00:02.000
              <span
                class="text-muted"
              >
                (2s)
              </span>
            </td>
          </tr>
          <tr>
            <td>
              Server conn. TCP handshake:
            </td>
            <td>
              1999-12-31 23:00:03.000
              <span
                class="text-muted"
              >
                (3s)
              </span>
            </td>
          </tr>
          <tr>
            <td>
              Response complete:
            </td>
            <td>
              1999-12-31 23:00:03.000
              <span
                class="text-muted"
              >
                (3s)
              </span>
            </td>
          </tr>
          <tr>
            <td>
              Server conn. TLS handshake:
            </td>
            <td>
              1999-12-31 23:00:04.000
              <span
                class="text-muted"
              >
                (4s)
              </span>
            </td>
          </tr>
          <tr>
            <td>
              Server conn. closed:
            </td>
            <td>
              1999-12-31 23:00:05.000
              <span
                class="text-muted"
              >
                (5s)
              </span>
            </td>
          </tr>
          <tr>
            <td>
              Client conn. closed:
            </td>
            <td>
              1999-12-31 23:00:06.000
              <span
                class="text-muted"
              >
                (6s)
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </section>
  </div>
</DocumentFragment>
`;

exports[`FlowView 6`] = `
<DocumentFragment>
  <div
    class="flow-detail"
  >
    <nav
      class="nav-tabs nav-tabs-sm"
    >
      <a
        class=""
        href="#"
      >
        Request
      </a>
      <a
        class=""
        href="#"
      >
        Response
      </a>
      <a
        class=""
        href="#"
      >
        WebSocket
      </a>
      <a
        class="active"
        href="#"
      >
        Error
      </a>
      <a
        class=""
        href="#"
      >
        Connection
      </a>
      <a
        class=""
        href="#"
      >
        Timing
      </a>
    </nav>
    <section
      class="error"
    >
      <div
        class="alert alert-warning"
      >
        error
        <div>
          <small>
            1999-12-31 23:00:07.000
          </small>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;

exports[`FlowView 7`] = `
<DocumentFragment>
  <div
    class="flow-detail"
  >
    <nav
      class="nav-tabs nav-tabs-sm"
    >
      <a
        class="active"
        href="#"
      >
        TCP Messages
      </a>
      <a
        class=""
        href="#"
      >
        Error
      </a>
      <a
        class=""
        href="#"
      >
        Connection
      </a>
      <a
        class=""
        href="#"
      >
        Timing
      </a>
    </nav>
    <section
      class="tcp"
    >
      <h4>
        TCP Data
      </h4>
      <div
        class="contentview"
      >
        <div
          class="controls"
        >
          <h5>
            2 Messages
          </h5>
          <a
            class="btn btn-default btn-xs"
            href="#"
          >
            <span>
              <i
                class="fa fa-fw fa-files-o"
              />
               
              <b>
                View:
              </b>
               auto 
              <span
                class="caret"
              />
            </span>
          </a>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;

exports[`FlowView 8`] = `
<DocumentFragment>
  <div
    class="flow-detail"
  >
    <nav
      class="nav-tabs nav-tabs-sm"
    >
      <a
        class=""
        href="#"
      >
        TCP Messages
      </a>
      <a
        class="active"
        href="#"
      >
        Error
      </a>
      <a
        class=""
        href="#"
      >
        Connection
      </a>
      <a
        class=""
        href="#"
      >
        Timing
      </a>
    </nav>
    <section
      class="error"
    >
      <div
        class="alert alert-warning"
      >
        error
        <div>
          <small>
            1999-12-31 23:00:07.000
          </small>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;

exports[`FlowView 9`] = `
<DocumentFragment>
  <div
    class="flow-detail"
  >
    <nav
      class="nav-tabs nav-tabs-sm"
    >
      <a
        class="active"
        href="#"
      >
        Request
      </a>
      <a
        class=""
        href="#"
      >
        Response
      </a>
      <a
        class=""
        href="#"
      >
        Error
      </a>
      <a
        class=""
        href="#"
      >
        Connection
      </a>
      <a
        class=""
        href="#"
      >
        Timing
      </a>
    </nav>
    <section
      class="dns-request"
    >
      <div
        class="first-line request-line"
      >
        <div>
          QUERY 
        </div>
      </div>
      <h5>
        Recursive Question
      </h5>
      <table>
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Type
            </th>
            <th>
              Class
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              dns.google
            </td>
            <td>
              A
            </td>
            <td>
              IN
            </td>
          </tr>
        </tbody>
      </table>
      <hr />
      <h5>
        Answer
      </h5>
      —
      <hr />
      <h5>
        Authority
      </h5>
      —
      <hr />
      <h5>
        Additional
      </h5>
      —
    </section>
  </div>
</DocumentFragment>
`;

exports[`FlowView 10`] = `
<DocumentFragment>
  <div
    class="flow-detail"
  >
    <nav
      class="nav-tabs nav-tabs-sm"
    >
      <a
        class=""
        href="#"
      >
        Request
      </a>
      <a
        class="active"
        href="#"
      >
        Response
      </a>
      <a
        class=""
        href="#"
      >
        Error
      </a>
      <a
        class=""
        href="#"
      >
        Connection
      </a>
      <a
        class=""
        href="#"
      >
        Timing
      </a>
    </nav>
    <section
      class="dns-response"
    >
      <div
        class="first-line response-line"
      >
        <div>
          NOERROR 
        </div>
      </div>
      <h5>
        Recursive Question
      </h5>
      <table>
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Type
            </th>
            <th>
              Class
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              dns.google
            </td>
            <td>
              A
            </td>
            <td>
              IN
            </td>
          </tr>
        </tbody>
      </table>
      <hr />
      <h5>
        Recursive Answer
      </h5>
      <table>
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Type
            </th>
            <th>
              Class
            </th>
            <th>
              TTL
            </th>
            <th>
              Data
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              dns.google
            </td>
            <td>
              A
            </td>
            <td>
              IN
            </td>
            <td>
              32
            </td>
            <td>
              8.8.8.8
            </td>
          </tr>
          <tr>
            <td>
              dns.google
            </td>
            <td>
              A
            </td>
            <td>
              IN
            </td>
            <td>
              32
            </td>
            <td>
              8.8.4.4
            </td>
          </tr>
        </tbody>
      </table>
      <hr />
      <h5>
        Authority
      </h5>
      —
      <hr />
      <h5>
        Additional
      </h5>
      —
    </section>
  </div>
</DocumentFragment>
`;

exports[`FlowView 11`] = `
<DocumentFragment>
  <div
    class="flow-detail"
  >
    <nav
      class="nav-tabs nav-tabs-sm"
    >
      <a
        class=""
        href="#"
      >
        Request
      </a>
      <a
        class=""
        href="#"
      >
        Response
      </a>
      <a
        class="active"
        href="#"
      >
        Error
      </a>
      <a
        class=""
        href="#"
      >
        Connection
      </a>
      <a
        class=""
        href="#"
      >
        Timing
      </a>
    </nav>
    <section
      class="error"
    >
      <div
        class="alert alert-warning"
      >
        error
        <div>
          <small>
            1999-12-31 23:00:07.000
          </small>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;
